src/Database/LSMTree/Internal/BlobFile.hs

```
@@ -51,24 +52,35 @@ instance NFData BlobSpan where
 
 -- | Open the given file to make a 'BlobFile'. The finaliser will close and
 -- delete the file.
+--
+-- REF: the resulting reference must be released once it is no longer used.
+--
+-- ASYNC: this should be called with asynchronous exceptions masked.
 {-# SPECIALISE openBlobFile :: HasCallStack => HasFS IO h -> FS.FsPath -> FS.OpenMode -> IO (Ref (BlobFile IO h)) #-}
 openBlobFile ::
-     PrimMonad m
+     (PrimMonad m, MonadCatch m)
   => HasCallStack
   => HasFS m h
   -> FS.FsPath
   -> FS.OpenMode
   -> m (Ref (BlobFile m h))
-openBlobFile fs path mode = do
-    blobFileHandle <- FS.hOpen fs path mode
-    let finaliser = do
-          FS.hClose fs blobFileHandle
-          FS.removeFile fs (FS.handlePath blobFileHandle)
-    newRef finaliser $ \blobFileRefCounter ->
-      BlobFile {
-        blobFileHandle,
-        blobFileRefCounter
-      }
+openBlobFile fs path mode =
+    bracketOnError (FS.hOpen fs path mode) (FS.hClose fs) $ \blobFileHandle -> do
```

Nit-picky: it's not clear to me that backetOnError is strictly needed here. If we assume async exceptions are masked then there nothing else within the body that would be a synchronisation point where an async exception could be delivered. And if we don't assume async exceptions are masked, then I think the code using `bracketOnError` means it works fine even without async exceptions being masked (but I agree callers still need to mask exceptions because they're allocating a resource).

---

> Nit-picky: it's not clear to me that backetOnError is strictly needed here. If we assume async exceptions are masked then there nothing else within the body that would be a synchronisation point where an async exception could be delivered.

Right, I'm not sure my reasoning for using `bracketOnError` here, but it doesn't hurt. I agree no async errors can currently be thrown, not sure if there are synchronous errors that can currently be thrown by the body

> And if we don't assume async exceptions are masked, then I think the code using `bracketOnError` means it works fine even without async exceptions being masked (but I agree callers still need to mask exceptions because they're allocating a resource).

Does it? As soon as we leave the scope of `bracketOnError` and we're in an unmasked state, then an async exception can be thrown before the caller gets to do something with the value, like putting it somewhere such that it will be cleaned up correctly